### PR TITLE
Bugfixes/shypike

### PR DESCRIPTION
**Pull request by shypike**
This should solve a vulnerability, work-around database locks and prevent failed saving of NZO files.

### CODE REVIEW

**Comment by shypike** [comment]
Should solve:
https://forums.sabnzbd.org/viewtopic.php?f=2&t=21173
https://github.com/sabnzbd/sabnzbd/issues/634
https://github.com/sabnzbd/sabnzbd/issues/636

**Comment by Safihre** [comment]
:+1: running smooth so far..

**Comment by Safihre** [comment]
Something is wrong with the changes.
When I download a post that is clearly gone from the servers, [like this one](https://github.com/sabnzbd/sabnzbd/files/406434/The.Big.Bang.Theory.S09E18.1080p.HDTV.X264-DIMENSION.nzb.gz), and ```prospective_par2``` is turned on, I can see in the console that it never discards articles because they are missing from all servers. So it will just go on forever at 50KB/s.. 
Turning off ```prospective_par2``` will make it discard articles again.

**But:**
When I switch back to my ```develop``` from #643  (also with ```prospective_par2``` turned on), articles are nicely discarded when both of my servers are tried and then it stops the download when its hopeless.

**Comment by shypike** [comment]
OK, back to the drawing board.

File: sabnzbd/decoder.py [hunk 1]
```diff
@@ -187,36 +187,9 @@ def run(self):
...
-                                    new_nzf = extrapars_sorted.pop()
-                                    nzo.add_parfile(new_nzf)
-                                    nzo.extrapars[parset] = extrapars_sorted
-                                    blocks_already = blocks_already + int_conv(new_nzf.blocks)
-                                    logging.info('Prospectively added %s repair blocks to %s', new_nzf.blocks, nzo.final_name)
-                                    # Reset all try lists
-                                    NzbQueue.do.reset_all_try_lists()
+                    nzo.prospecive_add()
+                    # Reset all try lists
+                    NzbQueue.do.reset_all_try_lists()
```
Comment (Safihre) [inline sabnzbd/decoder.py]: This resets the TryList after *every* missing article. 
So it needs to go...

File: sabnzbd/nzbstuff.py [hunk 2]
```diff
@@ -1185,6 +1193,41 @@ def remove_extrapar(self, parfile):
...
+                if parset in nzf.filename and self.extrapars[parset]:
+                    extrapars_sorted = sorted(self.extrapars[parset], key=lambda x: x.blocks, reverse=True)
+                    # Loop untill we have enough
+                    while blocks_already < total_need and extrapars_sorted:
+                        # Add the first one
+                        new_nzf = extrapars_sorted.pop()
+                        self.add_parfile(new_nzf)
+                        self.extrapars[parset] = extrapars_sorted
+                        blocks_already = blocks_already + int_conv(new_nzf.blocks)
+                        logging.info('Prospectively added %s repair blocks to %s', new_nzf.blocks, self.final_name)
```
Comment (Safihre) [inline sabnzbd/nzbstuff.py]: Here!
Only when a prospective block is added, we reset all.

I think we could limit the resetting to only when it's a ```.par2``` file, but not sure enough to change it now before the final.

File: sabnzbd/decoder.py (hunk 1 above)
Comment (shypike) [inline sabnzbd/decoder.py]: Case of bad refactoring.
I'll fix that.

**Comment by shypike** [comment]
I pushed a fix.
Should the reset be inside or outside the while-loop.
Inside looks like overkill to me, but could there be situations where the while-loop would run zero times?

**Comment by Safihre** [comment]
This will return False when there are no more par2-files left, right?
```
if parset in nzf.filename and self.extrapars[parset]
```

In that case, the location outside the loop would be best.

**Pull request merged**